numpy/_core/src/multiarray/stringdtype/static_string.c

```diff
@@ -0,0 +1,803 @@
+{
+    npy_string_allocator *allocator = m(sizeof(npy_string_allocator));
+    if (allocator == NULL) {
+        return NULL;
+    }
+    PyThread_type_lock *allocator_lock = PyThread_allocate_lock();
+    if (allocator_lock == NULL) {
+        f(allocator);
+        PyErr_SetString(PyExc_MemoryError, "Unable to allocate thread lock");
+        return NULL;
+    }
+    allocator->malloc = m;
+    allocator->free = f;
+    allocator->realloc = r;
+    // arena buffer gets allocated in arena_malloc
+    allocator->arena = NEW_ARENA;
+    allocator->allocator_lock = allocator_lock;
+
+    return allocator;
+}
+
+NPY_NO_EXPORT void
+NpyString_free_allocator(npy_string_allocator *allocator)
+{
+    npy_string_free_func f = allocator->free;
+
+    if (allocator->arena.buffer != NULL) {
+        f(allocator->arena.buffer);
+    }
+    if (allocator->allocator_lock != NULL) {
+        PyThread_free_lock(allocator->allocator_lock);
+    }
+
+    f(allocator);
+}
+
+static int
+is_short_string(const npy_packed_static_string *s)
+{
+    unsigned char high_byte =
+            ((_npy_static_string_u *)s)->direct_buffer.size_and_flags;
+    int has_short_flag = (high_byte & NPY_STRING_SHORT);
+    int has_on_heap_flag = (high_byte & NPY_STRING_ON_HEAP);
+    return has_short_flag && !has_on_heap_flag;
+}
+
+static int
+is_medium_string(const _npy_static_string_u *s)
+{
+    unsigned char high_byte = s->direct_buffer.size_and_flags;
+    int has_short_flag = (high_byte & NPY_STRING_SHORT);
+    int has_medium_flag = (high_byte & NPY_STRING_MEDIUM);
+    return (!has_short_flag && has_medium_flag);
+}
+
+NPY_NO_EXPORT int
+NpyString_isnull(const npy_packed_static_string *s)
+{
+    unsigned char high_byte =
+            ((_npy_static_string_u *)s)->direct_buffer.size_and_flags;
+    return (high_byte & NPY_STRING_MISSING) == NPY_STRING_MISSING;
+}
+
+static int
+is_not_a_vstring(const npy_packed_static_string *s)
+{
+    return is_short_string(s) || NpyString_isnull(s);
+}
+
+static int
+is_a_vstring(const npy_packed_static_string *s)
+{
+    return !is_not_a_vstring(s);
+}
+
+/*NUMPY_API
+ * Acquire the mutex locking the allocator attached to *descr*.
+ */
+npy_string_allocator *
+NpyString_acquire_allocator(PyArray_StringDTypeObject *descr)
+{
+    if (!PyThread_acquire_lock(descr->allocator->allocator_lock, NOWAIT_LOCK)) {
+        PyThread_acquire_lock(descr->allocator->allocator_lock, WAIT_LOCK);
+    }
+    return descr->allocator;
+}
+
+/*
+ * Simultaneously acquire the mutexes locking the allocators attached to
+ * *N* descriptors. The descriptors and resulting allocators must be interleaved.
+ * If pointers to the same descriptor are passed multiple times, only acquires
+ * the allocator mutex once.
+ *
+ * This function does a dynamic stack allocation and must not be called with
+ * arbitrarily large N.
+ *
+ * It would be nice to make this part of the public API, but right now varargs
+ * functions aren't supported in the API codegen.
```

Comment:
@seberg the `replace` ufunc needs 4 allocators, which breaks my assumption that people would need a maximum of 3 of them...

I tried to define varargs functions to replace the `allocator2` and `allocator3` variants I'm adding to the API in this PR as you suggested a while back but it turns out the C API codegen chokes trying to parse a varargs function, and indeed there aren't any varargs functions currently in the C API.

Does any of that sound familiar? Do you think it's worth diving into the codegen to try to add support for exposing a varargs function?

Reply:
It also looks like MSVC doesn't like the dynamically defined arrays, although I thought we specified we need C99...

Reply:
no VLAs can't be used... We could have a `allocators(nin, *dtypes)` function?  Although unless we skip non-strings there, it won't be super convenient to call generally.

Reply:
Ah good point about just letting people pass a pointer array, will update with that.